File: solutions/lua/protein-translation/1/protein-translation.lua
local function proteins(strand)
  local result = {}
  local codon_map = {
    Methionine = {"AUG"},
    Phenylalanine = {"UUU", "UUC"},
    Leucine = {"UUA", "UUG"},
    Serine = {"UCU", "UCC", "UCA", "UCG"},
    Tyrosine = {"UAU", "UAC"},
    Cysteine = {"UGU", "UGC"},
    Tryptophan = {"UGG"},
    STOP = {"UAA", "UAG", "UGA"}
  }
  local i = 1
  while i <= #strand do
    local s = strand:sub(i, i + 2)
    local found
    for acid, codons in pairs(codon_map) do
      for _, codon in ipairs(codons) do
        if s == codon then
          found = acid
          break
        end
      end
    end
    if found == nil then error()
    elseif found == 'STOP' then break
    else
      table.insert(result, found)
    end
    i = i + 3
  end
  return result
end

return { proteins = proteins }

File: solutions/go/protein-translation/1/protein_translation.go
package protein

import "errors"

func FromRNA(rna string) ([]string, error) {
	result := make([]string, len(rna)/3)
	for i := 0; i < len(rna); i += 3 {
		acid, err := FromCodon(rna[i : i+3])
		if err != nil {
			return []string{}, errors.New("")
		}
		result[i] = acid
	}
	return result, nil
}

func FromCodon(codon string) (string, error) {
	m := map[string]string{
		"AUG": "Methionine",
		"UUU": "Phenylalanine",
		"UUC": "Phenylalanine",
		"UUA": "Leucine",
		"UUG": "Leucine",
		"UCU": "Serine",
		"UCC": "Serine",
		"UCA": "Serine",
		"UCG": "Serine",
		"UAU": "Tyrosine",
		"UAC": "Tyrosine",
		"UGU": "Cysteine",
		"UGC": "Cysteine",
		"UGG": "Tryptophan",
		"UAA": "STOP",
		"UAG": "STOP",
		"UGA": "STOP",
	}
	v, ok := m[codon]
	if !ok {
		return "", errors.New("")
	}
	return v, nil
}
